### ISSUE TITLE
unnecessary_operation: reduces array of generics to just generic / compiler error

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#9951]
### Summary

.

### Lint Name

unnecessary_operation

### Reproducer

I tried this code:

```rust
#[derive(Clone, Default)]
struct MaybeCopy<T>(T);

impl Copy for MaybeCopy<u8> {}

fn main() {
    [MaybeCopy::default(); 13];
}
```

Clippy will try to make this
````rust
fn main() {
    MaybeCopy::default();
}
````
for some reason, maybe we could just remove the entire array though? :thinking: 

I saw this happen:

```
This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0282]: type annotations needed
 --> src/main.rs:7:5
  |
7 |     MaybeCopy::default();
  |     ^^^^^^^^^ cannot infer type for type parameter `T` declared on the struct `MaybeCopy`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0282`.
Original diagnostics will follow.

warning: unnecessary operation
 --> src/main.rs:7:5
  |
7 |     [MaybeCopy::default(); 13];
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: statement can be reduced to: `MaybeCopy::default();`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_operation
  = note: `#[warn(clippy::unnecessary_operation)]` on by default
```

### Version

```text
rustc 1.67.0-nightly (b3bc6bf31 2022-11-24)
binary: rustc
commit-hash: b3bc6bf31265ac10946a0832092dbcedf9b26805
commit-date: 2022-11-24
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```

**Comment by Jarcho** [comment]
`MaybeCopy::default()` could have side-effects, so it still needs to be called.

Also, wow. The compiler is really trigger happy on type deduction there.

**Comment by AloeareV** [comment]
Recently hit this on stable.
